### ISSUE TITLE
Using wesnoth.sides.create() causes the client to segfault

### ISSUE THREAD

**Issue by Pentarctagon** [wesnoth/wesnoth#6230]
Steps taken:
* add the code
```
[event]
    name=turn 1
    [lua]
        code=<< wesnoth.sides.create() >>
    [/lua]
[/event]
```
to Aethermaw's scenario file.
* Connect to wesnothd.
* Start a game with myself controlling both sides.
* Wesnoth segfaults.

**Comment by Pentarctagon** [comment]
Note also that if/when this is fixed, wesnothd should also be updated to insert a row into the database when a side is created so that information is kept accurate.

**Comment by gfgtdf** [comment]
I didn't look i to this deeply but the cause is probably that when the teams are stored as a "std::vector<team>", then adding a team invalidates all present references to team objects, there are basically 3 solution if we want to fix this:
1) store teams as vector<u ique_ptr<team>>
2) remove all codes that store references to team objects (and always use indices
3) don't allow create_side to be called in contexts where code hold references to team objects (maybe it will work i.n preload event, not sure)


> Note also that if/when this is fixed, wesnothd should also be updated to insert a row into the database when a side is created so that information is kept accurate.


Wesnothd does already add a row to the game::sides_ object, not sure about the database

**Comment by CelticMinstrel** [comment]
I have an even better solution. If you simply store them as a `std::deque<team>` instead, then adding a new one to the end won't invalidate references to all the others.

**Comment by soliton-** [comment]
IMO it'd be better to not support dynamic side creation.

**Comment by Pentarctagon** [comment]
I mean, if the feature is entirely unusable and has been since it was introduced then it may be "in" in the most general sense, but there's also nothing that would be impacted by removing it.

**Comment by gfgtdf** [comment]
An alternative to support the ist usecase described above would be to allow modification to contain [side] but the problem with that approach of mis that the modification then does not know which numbert the side has

**Comment by CelticMinstrel** [comment]
> I mean, if the feature is entirely unusable and has been since it was introduced then it may be "in" in the most general sense, but there's also nothing that would be impacted by removing it.

I don't recall trying it myself, but… I don't think it's likely that it has always been unusable? Wasn't it at least tested back when it was first implemented? I wasn't the one to do that though so I can't say for sure.

**Comment by Pentarctagon** [comment]
Checking at 45a33e6e3aafc2bb70fb1249d347429ca54eb212, the segfault always happens for a `start` event. In a moveto, it seems to somehow be related to the faction chosen (???) - if I choose Drakes, Knalgans, or Rebels then it doesn't segfault (but still behaves wrongly in other ways such as making the whole map fogged and then removing all fog if I open and close the side list), but if I choose any other Default faction then it doesn't segfault.

**Comment by gfgtdf** [comment]
Hmm i was sure i tested it a bit, but then of course such crashes can always depend on random factors like compiler/std used etc. And i was also aware that is was not fully reliable yet.

**Comment by CelticMinstrel** [comment]
Maybe we could remove it, technically, but I think it's nice to have so I don't really want to remove it.

Has anyone tried my suggestion to fix this crash?

**Comment by Pentarctagon** [comment]
I haven't, at least. I think my preference would be to remove the current implementation (closing this issue) and then open another issue to discuss re-adding it. There are almost certainly going to be other issues with making it possible to increase the number of sides beyond just this segfault, and the API itself could also be improved. In particular, it would be good to be able to specify a side definition rather than simply creating an empty side and then needing to set it up afterwards.

**Comment by CelticMinstrel** [comment]
I don't really like the idea of "removing with intent to re-add". If that's what we're doing, in my opinion it should all be within a single pull request.

**Comment by soliton-** [comment]
If you want to keep it then what are the use cases? How should other tools deal with this? For example what should be shown in the multiplayer lobby for a game with a dynamic amount of sides? How should game info for such a game be recorded in the database? Is the assumption that those sides are never player controlled and some helper sides that can be ignored?

**Comment by CelticMinstrel** [comment]
> If you want to keep it then what are the use cases?

Maybe whenever you capture a village it spawns mutually hostile "barbarians" which will attack you but may also attack each other.

> How should other tools deal with this?

That question is too vague to answer. It depends on the tool.

> For example what should be shown in the multiplayer lobby for a game with a dynamic amount of sides?

I think I'd just show the current number of sides. Although generally speaking it might be more useful to instead show the number of human-controlled sides, rather than the total number of sides.

> How should game info for such a game be recorded in the database?

Basically the same as above – record the number of sides as of scenario end, or just record the number of human-controlled sides.

> Is the assumption that those sides are never player controlled and some helper sides that can be ignored?

I can't think of a use-case where a dynamic side is player-controlled, so I think this is _probably_ a safe assumption.

**Comment by gfgtdf** [comment]
I just noticed, that there is even a wml test for this api : `test_create_side`

**Comment by gfgtdf** [comment]
> and the API itself could also be improved. In particular, it would be good to be able to specify a side definition rather than simply creating an empty side and then needing to set it up afterwards.

It does accept a wml table iirc, except controller= and side= all keys should be supported.

**Comment by gfgtdf** [comment]
> Steps taken: ...

I tested and i got a different behaviour: the map show as normal but tiles that the cursor moves over become black, but i guess with ub everything is possible

**Comment by gfgtdf** [comment]
For the record, the server code to handle the create side notification is currently broken (it doesn't send feedback to the clients so that they'll wait forever), but of course that shouldn't make the Client segfault.Even though that is an easy fix, I think we should fix the client first.

**Comment by soliton-** [comment]
> > How should other tools deal with this?
> 
> That question is too vague to answer. It depends on the tool.

Seems pretty clear to me. Every tool will have to implement the same logic as wesnoth. So every tool will need to parse and understand at a minimum the side creation tag. The more such tags a tool has to implement the smaller the chance we'll get good tools. If we're going to make wesnoth do everything like the preprocessing for example that's cool. But that's not always practical.
 
> > Is the assumption that those sides are never player controlled and some helper sides that can be ignored?
> 
> I can't think of a use-case where a dynamic side is player-controlled, so I think this is _probably_ a safe assumption.

Unless that is enforced in any way that doesn't help much. If it was enforced so that you could ignore the dynamic side stuff in a lot of cases it might make this somewhat sane. Still a hard sell IMO. Lots of work so that you can implement hypothetically infinite barbarian sides.
